src/go/wsl-helper/pkg/wsl-utils/version_windows.go

```diff
@@ -232,85 +172,9 @@ func (v PackageVersion) Less(other PackageVersion) bool {
 	return false
 }
 
-// packageInfo corresponds to the PACKAGE_INFO structure.
-type packageInfo struct {
-	reserved          uint32
-	flags             uint32
-	path              *uint16
-	packageFullName   *uint16
-	packageFamilyName *uint16
-	packageID         struct {
-		reserved              uint32
-		processorArchitecture uint32
-		version               PackageVersion
-		name                  *uint16
-		publisher             *uint16
-		resourceID            *uint16
-		publisherID           *uint16
-	}
-}
-
-// getPackageVersion gets the package version of the package with the given
-// full name.
-func getPackageVersion(fullName string) (*PackageVersion, error) {
-	nameBuffer, err := windows.UTF16PtrFromString(fullName)
-	if err != nil {
-		return nil, err
-	}
-	var packageInfoReference uintptr
-	rv, _, err := openPackageInfoByFullName.Call(
-		uintptr(unsafe.Pointer(nameBuffer)),
-		0, // reserved
-		uintptr(unsafe.Pointer(&packageInfoReference)),
-	)
-	if rv != uintptr(windows.ERROR_SUCCESS) {
-		return nil, errorFromWin32("error opening package info", rv, err)
-	}
-	defer func() { _, _, _ = closePackageInfo.Call(packageInfoReference) }()
-
-	var bufferLength, count uint32
-	rv, _, err = getPackageInfo.Call(
-		packageInfoReference,
-		uintptr(PACKAGE_INFORMATION_BASIC|PACKAGE_FILTER_STATIC|PACKAGE_FILTER_DYNAMIC),
-		uintptr(unsafe.Pointer(&bufferLength)),
-		uintptr(unsafe.Pointer(nil)),
-		uintptr(unsafe.Pointer(nil)),
-	)
-	switch rv {
-	case uintptr(windows.ERROR_SUCCESS):
-		break
-	case uintptr(windows.ERROR_INSUFFICIENT_BUFFER):
-		// This is expected: we didn't provide any buffer
-		break
-	default:
-		return nil, errorFromWin32("error getting buffer size", rv, err)
-	}
-
-	buf := make([]byte, bufferLength)
-	rv, _, err = getPackageInfo.Call(
-		packageInfoReference,
-		uintptr(PACKAGE_INFORMATION_BASIC|PACKAGE_FILTER_STATIC|PACKAGE_FILTER_DYNAMIC),
-		uintptr(unsafe.Pointer(&bufferLength)),
-		uintptr(unsafe.Pointer(unsafe.SliceData(buf))),
-		uintptr(unsafe.Pointer(&count)),
-	)
-	if rv != uintptr(windows.ERROR_SUCCESS) {
-		return nil, errorFromWin32("error getting package info", rv, err)
-	}
-	infos := unsafe.Slice((*packageInfo)(unsafe.Pointer(unsafe.SliceData(buf))), count)
-	for _, info := range infos {
-		// `info` is a pointer to an unsafe slice; make a copy of the version
-		// on the stack and then return that so the GC knows about it.
-		versionCopy := info.packageID.version
-		return &versionCopy, nil
-	}
-
-	return nil, fmt.Errorf("no info found for %s", fullName)
-}
-
-// Get the component versions for an AppX-based installation.  Returns the WSL
+// Get the component versions by asking the CLI.  Returns the WSL
 // version, followed by the kernel version.
-func getAppxVersion(ctx context.Context, log *logrus.Entry) (*PackageVersion, *PackageVersion, error) {
+func getVersionFromCLI(ctx context.Context, log *logrus.Entry) (*PackageVersion, *PackageVersion, error) {
```

Comment:
I find it a bit weird that the function returns 2 pointers instead of 2 structs, but I guess the old version was already like that, and you don't want to change it?

Reply:
Yeah, I didn't see the point of changing it. Also, returning the pointers means passing slightly less stuff around (one of the linters we haven't enabled yet complains about passing structs by value).

I originally didn't think this existed and would have made a function to just return a `*WSLInfo` instead, but this was already there…

Reply:
> Also, returning the pointers means passing slightly less stuff around (one of the linters we haven't enabled yet complains about passing structs by value).

This is actually a myth for small structures, as cache lines are typically 64 bytes, so passing the struct is only a single cache hit, whereas using a pointer will require 2 cache accesses. So passing small structs by value is typically better.

But this rarely matters. I find it more important that when you return a pointer, the caller should always check if the returned pointer is `nil` before dereferencing it. If the pointer can't be `nil`, just pass the value and don't worry about it.